addons/sale_expense/models/hr_expense_sheet.py

```diff
@@ -15,42 +16,54 @@ def _compute_sale_order_count(self):
     def _get_sale_order_lines(self):
         """
             This method is used to try to find the sale order lines created by expense sheets.
-
-            :return: sale.order.line
-            :rtype: recordset
+            It is used to reset the quantities of the sale order lines when the expense sheet is reset.
+            It uses several shared fields to try to find the sale order lines:
+                - order_id
+                - product_id
+                - product_uom_qty
+                - sale order line's price_unit (computed from the product_id, then rounded to the currency's rounding)
+                - name
         """
-        expensed_amls = self.account_move_id.line_ids.filtered(lambda aml: aml.expense_id.sale_order_id and aml.balance >= 0)
+        # Get the product account move lines created by an expense
+        expensed_amls = self.account_move_id.line_ids.filtered(lambda aml: aml.expense_id.sale_order_id and aml.balance >= 0 and not aml.tax_line_id)
         if not expensed_amls:
             return self.env['sale.order.line']
+
+        # Get the sale orders linked to the related expenses
         aml_to_so_map = expensed_amls._sale_determine_order()
-        sale_order_ids = tuple(set(aml_to_so_map[aml.id].id for aml in expensed_amls))
-        aml_sol_unit_price_map = dict(expensed_amls.mapped(lambda aml: (aml.id, aml._sale_get_invoice_price(aml_to_so_map[aml.id]))))
-        product_ids = tuple(expensed_amls.product_id.ids)
-        quantities = tuple(expensed_amls.mapped('quantity'))
-        names = tuple(expensed_amls.mapped('name'))
+
         self.env['sale.order.line'].flush_model(['order_id', 'product_id', 'product_uom_qty', 'price_unit', 'name'])
+        self.env['res.company'].flush_model(['currency_id'])
+        self.env['res.currency'].flush_model(['rounding'])
         query = """
-            SELECT 
-                DISTINCT ON (sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name)
-                sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name, sol.id
-            FROM sale_order_line AS sol
-            WHERE sol.is_expense = TRUE
-                AND sol.order_id IN %s
-                AND sol.product_id IN %s
-                AND sol.product_uom_qty IN %s
-                AND sol.price_unit IN %s
-                AND sol.name IN %s
-            ORDER BY sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name
+              WITH aml(id, order_id, product_id, product_uom_qty, price_unit, name) AS (VALUES %s)
```

Comment:
Looks like the `id`/`expense_id` is not used

```suggestion
              WITH aml(order_id, product_id, product_uom_qty, price_unit, name) AS (VALUES %s)
```